src/cata_ui.cpp

```diff
@@ -0,0 +1,1038 @@
+        return false;
+    }
+    return uncovered[y * get_rect().size_x + x];
+}
+
+long auto_bordered_window::get_border_char( unsigned int x, unsigned int y ) const
+{
+    bool left = is_uncovered( x - 1, y );
+    bool up = is_uncovered( x, y - 1 );
+    bool right = is_uncovered( x + 1, y );
+    bool down = is_uncovered( x, y + 1 );
+
+    long ret = ' ';
+
+    if( left ) {
+        ret = LINE_OXOX; // '-'
+    }
+
+    if( up ) {
+        if( left ) {
+            ret = LINE_XOOX; // '_|'
+        } else {
+            ret = LINE_XOXO; // '|'
+        }
+    }
+
+    if( right ) {
+        if( up ) {
+            if( left ) {
+                ret = LINE_XXOX; // '_|_'
+            } else {
+                ret = LINE_XXOO; // '|_'
+            }
+        } else {
+            ret = LINE_OXOX; // '-'
+        }
+    }
+
+    if( down ) {
+        if( right ) {
+            if( up ) {
+                if( left ) {
+                    ret = LINE_XXXX; // '-|-'
+                } else {
+                    ret = LINE_XXXO; // '|-'
+                }
+            } else {
+                if( left ) {
+                    ret = LINE_OXXX; // '^|^'
+                } else {
+                    ret = LINE_OXXO; // '|^'
+                }
+            }
+        } else {
+            if( left ) {
+                if( up ) {
+                    ret = LINE_XOXX; // '-|'
+                } else {
+                    ret = LINE_OOXX; // '^|'
+                }
+            } else {
+                ret = LINE_XOXO; // '|'
+            }
+        }
+    }
+
+    return ret;
+}
+
+void auto_bordered_window::local_draw()
+{
+    ui_window::local_draw();
+
+    auto win = get_win(); // never null
+    for( unsigned int x = 0; x < get_rect().size_x; x++ ){
+        for( unsigned int y = 0; y < get_rect().size_y; y++ ) {
+            if( is_uncovered( x, y ) ) {
+                mvwputch( win, y, x, border_color, get_border_char( x, y ) );
+            }
+        }
+    }
+}
+
+ui_vertical_list::ui_vertical_list( size_t size_x, size_t size_y, int x, int y, ui_anchor anchor ) : ui_element( size_x, size_y, x, y, anchor )
+{
+}
+
+void ui_vertical_list::draw()
+{
+    auto win = get_win();
+    if( win == nullptr ) {
+        return;
+    }
+
+    unsigned int start_line = window_scroll;
+    unsigned int end_line = start_line + get_rect().size_y;
+
+    unsigned int available_space = get_rect().size_x - 2; // 2 for scroll bar and spacer
+
+    for( unsigned int line = start_line; line < end_line && line < text.size(); line++ ) {
+        auto txt = text[line];
+        if( txt.size() > available_space ) {
+            txt = txt.substr( 0, available_space );
+        }
+        if( scroll == line ) {
+            mvwprintz( win, get_ay() + line - start_line, get_ax() + 2, hilite(text_color), txt.c_str() );
+        } else {
+            mvwprintz( win, get_ay() + line - start_line, get_ax() + 2, text_color, txt.c_str() );
```

Comment:
Considering that `txt` may be anything, including format specifiers, it should be printed via `"%s", txt.c_str()`.